### Importing required libraries

In [1]:
import pandas as pd

### Importing Source Datafile for Pre-Processing

In [2]:
#df = pd.read_csv('cc_transaction.csv')

In [3]:
#df.head()

In [4]:
#df_sample=df.head(900000)

In [5]:
#saving the sample data

#df_sample.to_csv('credit_debit_transactions.csv')

In [6]:
#df=df_sample

In [7]:
#df=pd.read_csv('credit_debit_transactions.csv')
df=pd.read_csv('reduced_transaction.csv')

In [8]:
pd.set_option('display.max_columns', None)
df=df.drop(['Unnamed: 0.1','Unnamed: 0'], axis=1)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000000 entries, 0 to 4999999
Data columns (total 36 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   User                    int64  
 1   Card                    int64  
 2   Year                    int64  
 3   Month                   int64  
 4   Day                     int64  
 5   Time                    object 
 6   Amount                  object 
 7   Use Chip                object 
 8   Merchant Name           int64  
 9   Merchant City           object 
 10  Merchant State          object 
 11  Zip                     float64
 12  MCC                     int64  
 13  Errors?                 object 
 14  Is Fraud?               object 
 15  Person                  object 
 16  Current Age             int64  
 17  Birth Year              int64  
 18  Birth Month             int64  
 19  Gender                  object 
 20  Latitude                float64
 21  Longitude               float64

Columnns with Missing data:
    1. Merchant State
    2. Zip
    3. Error?
    4. user_Apartment

Among 44 columns there are only 4 columns with missing values.  But, 'user_apartment' may not be needed so dropping the user_apartment column

In [9]:
#df=df.drop('user_Apartment', axis=1)

In [10]:
#removing the $ sign from the Amount column
# columns=['Yearly Income - Person','Per Capita Income - Zipcode','Credit Limit','Amount']
# for col in columns:
#     df[col]=df[col].replace(r'[$]', '', regex=True)

Amount is supposed to be a numerical column, but the currency symbol in it makes it as a object which is not suitable to train a ML model.  So removed the currency symbol from the column

In [11]:
df.head()

User  Card  Year  Month  Day   Time   Amount           Use Chip  \
0     0     0  2002      9    1  06:21  $134.09  Swipe Transaction   
1     0     0  2002      9    1  06:42   $38.48  Swipe Transaction   
2     0     0  2002      9    2  06:22  $120.34  Swipe Transaction   
3     0     0  2002      9    2  17:45  $128.95  Swipe Transaction   
4     0     0  2002      9    3  06:23  $104.71  Swipe Transaction   

         Merchant Name  Merchant City Merchant State      Zip   MCC   Errors?  \
0  3527213246127876953       La Verne             CA  91750.0  5300  No Error   
1  -727612092139916043  Monterey Park             CA  91754.0  5411  No Error   
2  -727612092139916043  Monterey Park             CA  91754.0  5411  No Error   
3  3414527459579106770  Monterey Park             CA  91754.0  5651  No Error   
4  5817218446178736267       La Verne             CA  91750.0  5912  No Error   

  Is Fraud?          Person  Current Age  Birth Year  Birth Month  Gender  \
0        No  Hazel Robinson           53        1966           11  Female   
1        No  Hazel Robinson           53        1966           11  Female   
2        No  Hazel Robinson           53        1966           11  Female   
3        No  Hazel Robinson           53        1966           11  Female   
4        No  Hazel Robinson           53        1966           11  Female   

   Latitude  Longitude Yearly Income - Person  FICO Score  Num Credit Cards  \
0     34.15    -117.76                 $59696         787                 5   
1     34.15    -117.76                 $59696         787                 5   
2     34.15    -117.76                 $59696         787                 5   
3     34.15    -117.76                 $59696         787                 5   
4     34.15    -117.76                 $59696         787                 5   

  Card Brand Card Type       Card Number  Expires  CVV Has Chip  Cards Issued  \
0       Visa     Debit  4344676511950444  12/2022  623      YES             2   
1       Visa     Debit  4344676511950444  12/2022  623      YES             2   
2       Visa     Debit  4344676511950444  12/2022  623      YES             2   
3       Visa     Debit  4344676511950444  12/2022  623      YES             2   
4       Visa     Debit  4344676511950444  12/2022  623      YES             2   

  Credit Limit Acct Open Date  Year PIN last Changed Card on Dark Web  
0       $24295        09/2002                   2008               No  
1       $24295        09/2002                   2008               No  
2       $24295        09/2002                   2008               No  
3       $24295        09/2002                   2008               No  
4       $24295        09/2002                   2008               No

Removed all the currency symbols

It is observed that for the Online transactions we have missing 'Merchant State' and 'Zip'.  So for online transactions 'Merchant State' will be 'online' and 'Zip' will be '0000'

In [12]:
#df['Merchant State']=df['Merchant State'].fillna('online')

In [13]:
#df['Zip']=df['Zip'].fillna(0000)

In [14]:
#df['Errors?']=df['Errors?'].fillna('No Error')

In [15]:
#df.to_csv('credit_debit_preprocessed_transactions.csv')

In [16]:
df['Errors?']=df['Errors?'].apply(lambda x: 1 if x != 'No Error' else 0)

In [17]:
df['Errors?'].value_counts()

Errors?
0    4917515
1      82485
Name: count, dtype: int64

In [18]:
columns=['Yearly Income - Person','Credit Limit','Amount']
for col in columns:
    df[col]=df[col].replace(r'[$]', '', regex=True)

In [19]:
df.head()

User  Card  Year  Month  Day   Time  Amount           Use Chip  \
0     0     0  2002      9    1  06:21  134.09  Swipe Transaction   
1     0     0  2002      9    1  06:42   38.48  Swipe Transaction   
2     0     0  2002      9    2  06:22  120.34  Swipe Transaction   
3     0     0  2002      9    2  17:45  128.95  Swipe Transaction   
4     0     0  2002      9    3  06:23  104.71  Swipe Transaction   

         Merchant Name  Merchant City Merchant State      Zip   MCC  Errors?  \
0  3527213246127876953       La Verne             CA  91750.0  5300        0   
1  -727612092139916043  Monterey Park             CA  91754.0  5411        0   
2  -727612092139916043  Monterey Park             CA  91754.0  5411        0   
3  3414527459579106770  Monterey Park             CA  91754.0  5651        0   
4  5817218446178736267       La Verne             CA  91750.0  5912        0   

  Is Fraud?          Person  Current Age  Birth Year  Birth Month  Gender  \
0        No  Hazel Robinson           53        1966           11  Female   
1        No  Hazel Robinson           53        1966           11  Female   
2        No  Hazel Robinson           53        1966           11  Female   
3        No  Hazel Robinson           53        1966           11  Female   
4        No  Hazel Robinson           53        1966           11  Female   

   Latitude  Longitude Yearly Income - Person  FICO Score  Num Credit Cards  \
0     34.15    -117.76                  59696         787                 5   
1     34.15    -117.76                  59696         787                 5   
2     34.15    -117.76                  59696         787                 5   
3     34.15    -117.76                  59696         787                 5   
4     34.15    -117.76                  59696         787                 5   

  Card Brand Card Type       Card Number  Expires  CVV Has Chip  Cards Issued  \
0       Visa     Debit  4344676511950444  12/2022  623      YES             2   
1       Visa     Debit  4344676511950444  12/2022  623      YES             2   
2       Visa     Debit  4344676511950444  12/2022  623      YES             2   
3       Visa     Debit  4344676511950444  12/2022  623      YES             2   
4       Visa     Debit  4344676511950444  12/2022  623      YES             2   

  Credit Limit Acct Open Date  Year PIN last Changed Card on Dark Web  
0        24295        09/2002                   2008               No  
1        24295        09/2002                   2008               No  
2        24295        09/2002                   2008               No  
3        24295        09/2002                   2008               No  
4        24295        09/2002                   2008               No

In [20]:
df=df.drop('Person', axis=1)

In [21]:
df['Gender']=df['Gender'].apply(lambda x: 1 if x == 'Male' else 0)

In [22]:
df=df.drop(['Merchant City','Merchant State'], axis=1)

In [23]:
df['Use Chip'].value_counts()

Use Chip
Swipe Transaction     3118560
Chip Transaction      1272531
Online Transaction     608909
Name: count, dtype: int64

Use Chip
Swipe Transaction     311856  => Encoded as 10
Chip Transaction      12725  => Encoded as 231
Online Transaction     608  => Encoded as 3909

In [24]:
#df['Use Chip']=df['Use Chip'].apply(lambda x: 1 if x == 'Swipe Transaction' else (lambda x: 2 if x == 'Chip Transaction' else 3))

In [25]:
def time_to_minutes(time_str):
    hours, minutes = map(int, time_str.split(':'))
    return hours * 60 + minutes

df['Time']=df['Time'].apply(time_to_minutes)

In [26]:
df.head()

User  Card  Year  Month  Day  Time  Amount           Use Chip  \
0     0     0  2002      9    1   381  134.09  Swipe Transaction   
1     0     0  2002      9    1   402   38.48  Swipe Transaction   
2     0     0  2002      9    2   382  120.34  Swipe Transaction   
3     0     0  2002      9    2  1065  128.95  Swipe Transaction   
4     0     0  2002      9    3   383  104.71  Swipe Transaction   

         Merchant Name      Zip   MCC  Errors? Is Fraud?  Current Age  \
0  3527213246127876953  91750.0  5300        0        No           53   
1  -727612092139916043  91754.0  5411        0        No           53   
2  -727612092139916043  91754.0  5411        0        No           53   
3  3414527459579106770  91754.0  5651        0        No           53   
4  5817218446178736267  91750.0  5912        0        No           53   

   Birth Year  Birth Month  Gender  Latitude  Longitude  \
0        1966           11       0     34.15    -117.76   
1        1966           11       0     34.15    -117.76   
2        1966           11       0     34.15    -117.76   
3        1966           11       0     34.15    -117.76   
4        1966           11       0     34.15    -117.76   

  Yearly Income - Person  FICO Score  Num Credit Cards Card Brand Card Type  \
0                  59696         787                 5       Visa     Debit   
1                  59696         787                 5       Visa     Debit   
2                  59696         787                 5       Visa     Debit   
3                  59696         787                 5       Visa     Debit   
4                  59696         787                 5       Visa     Debit   

        Card Number  Expires  CVV Has Chip  Cards Issued Credit Limit  \
0  4344676511950444  12/2022  623      YES             2        24295   
1  4344676511950444  12/2022  623      YES             2        24295   
2  4344676511950444  12/2022  623      YES             2        24295   
3  4344676511950444  12/2022  623      YES             2        24295   
4  4344676511950444  12/2022  623      YES             2        24295   

  Acct Open Date  Year PIN last Changed Card on Dark Web  
0        09/2002                   2008               No  
1        09/2002                   2008               No  
2        09/2002                   2008               No  
3        09/2002                   2008               No  
4        09/2002                   2008               No

In [27]:
def year_month_to_int(year_month_col):
    year = year_month_col.str.split('/').str[1]
    month = year_month_col.str.split('/').str[0]
    return (year + month).astype(int)

df['Expires']=year_month_to_int(df['Expires'])

In [28]:
df['Acct Open Date']=year_month_to_int(df['Acct Open Date'])

In [29]:
df['Card Brand'].value_counts()

Card Brand
Mastercard    2822955
Visa          1795780
Amex           286657
Discover        94608
Name: count, dtype: int64

Mastercard    2822955 => Encoded as 1
Visa          1795780 => Encoded as 2
Amex           286657 => Encoded as 3
Discover        94608 => Encoded as 4

In [30]:
#df['Card Brand']=df['Card Brand'].apply(lambda x: 1 if x == 'Mastercard' else (lambda x: 2 if x == 'Visa' else (lambda x: 3 if x == 'Amex' else 4)))

In [31]:
df['Card Type'].value_counts()

Card Type
Debit              3160917
Credit             1508076
Debit (Prepaid)     331007
Name: count, dtype: int64

Debit              3160917 => Encoded as 1 ||

Credit             150807 => Encoded as 2 ||
6
Debit (Prepaid)     3310 => Encoded as 3 ||07

In [32]:
#df['Card Type']=df['Card Type'].apply(lambda x: 1 if x == 'Debit' else(lambda x: 2 if x == 'Credit' else 3))

In [33]:
df['Has Chip']=df['Has Chip'].apply(lambda x: 1 if x== 'YES' else 0)
df['Card on Dark Web']=df['Card on Dark Web'].apply(lambda x: 0 if x == 'No' else 1)

In [34]:
df.head()

User  Card  Year  Month  Day  Time  Amount           Use Chip  \
0     0     0  2002      9    1   381  134.09  Swipe Transaction   
1     0     0  2002      9    1   402   38.48  Swipe Transaction   
2     0     0  2002      9    2   382  120.34  Swipe Transaction   
3     0     0  2002      9    2  1065  128.95  Swipe Transaction   
4     0     0  2002      9    3   383  104.71  Swipe Transaction   

         Merchant Name      Zip   MCC  Errors? Is Fraud?  Current Age  \
0  3527213246127876953  91750.0  5300        0        No           53   
1  -727612092139916043  91754.0  5411        0        No           53   
2  -727612092139916043  91754.0  5411        0        No           53   
3  3414527459579106770  91754.0  5651        0        No           53   
4  5817218446178736267  91750.0  5912        0        No           53   

   Birth Year  Birth Month  Gender  Latitude  Longitude  \
0        1966           11       0     34.15    -117.76   
1        1966           11       0     34.15    -117.76   
2        1966           11       0     34.15    -117.76   
3        1966           11       0     34.15    -117.76   
4        1966           11       0     34.15    -117.76   

  Yearly Income - Person  FICO Score  Num Credit Cards Card Brand Card Type  \
0                  59696         787                 5       Visa     Debit   
1                  59696         787                 5       Visa     Debit   
2                  59696         787                 5       Visa     Debit   
3                  59696         787                 5       Visa     Debit   
4                  59696         787                 5       Visa     Debit   

        Card Number  Expires  CVV  Has Chip  Cards Issued Credit Limit  \
0  4344676511950444   202212  623         1             2        24295   
1  4344676511950444   202212  623         1             2        24295   
2  4344676511950444   202212  623         1             2        24295   
3  4344676511950444   202212  623         1             2        24295   
4  4344676511950444   202212  623         1             2        24295   

   Acct Open Date  Year PIN last Changed  Card on Dark Web  
0          200209                   2008                 0  
1          200209                   2008                 0  
2          200209                   2008                 0  
3          200209                   2008                 0  
4          200209                   2008                 0

In [36]:

def encode_card_brand(card_brand):
    card_brand_mapping = {
        'Mastercard': 1,
        'Visa': 2,
        'Amex': 3,
        'Discover': 4
    }

    return card_brand_mapping.get(card_brand, 0)

df['Card Brand'] = df['Card Brand'].apply(encode_card_brand)

In [38]:
def encode_use_chip(use_chip):
    use_chip_mapping = {
        'Swipe Transaction': 1,
        'Chip Transaction': 2,
        'Online Transaction': 3
    }

    return use_chip_mapping.get(use_chip, 0)

df['Use Chip'] = df['Use Chip'].apply(encode_use_chip)

In [41]:
def encode_card_type(card_type):
    card_type_mapping = {
        'Debit': 1,
        'Credit': 2,
        'Debit (Prepaid)': 3
    }

    return card_type_mapping.get(card_type, 0)

df['Card Type'] = df['Card Type'].apply(encode_card_type)

In [42]:
df.head()

User  Card  Year  Month  Day  Time  Amount  Use Chip        Merchant Name  \
0     0     0  2002      9    1   381  134.09         1  3527213246127876953   
1     0     0  2002      9    1   402   38.48         1  -727612092139916043   
2     0     0  2002      9    2   382  120.34         1  -727612092139916043   
3     0     0  2002      9    2  1065  128.95         1  3414527459579106770   
4     0     0  2002      9    3   383  104.71         1  5817218446178736267   

       Zip   MCC  Errors? Is Fraud?  Current Age  Birth Year  Birth Month  \
0  91750.0  5300        0        No           53        1966           11   
1  91754.0  5411        0        No           53        1966           11   
2  91754.0  5411        0        No           53        1966           11   
3  91754.0  5651        0        No           53        1966           11   
4  91750.0  5912        0        No           53        1966           11   

   Gender  Latitude  Longitude Yearly Income - Person  FICO Score  \
0       0     34.15    -117.76                  59696         787   
1       0     34.15    -117.76                  59696         787   
2       0     34.15    -117.76                  59696         787   
3       0     34.15    -117.76                  59696         787   
4       0     34.15    -117.76                  59696         787   

   Num Credit Cards  Card Brand  Card Type       Card Number  Expires  CVV  \
0                 5           2          1  4344676511950444   202212  623   
1                 5           2          1  4344676511950444   202212  623   
2                 5           2          1  4344676511950444   202212  623   
3                 5           2          1  4344676511950444   202212  623   
4                 5           2          1  4344676511950444   202212  623   

   Has Chip  Cards Issued Credit Limit  Acct Open Date  Year PIN last Changed  \
0         1             2        24295          200209                   2008   
1         1             2        24295          200209                   2008   
2         1             2        24295          200209                   2008   
3         1             2        24295          200209                   2008   
4         1             2        24295          200209                   2008   

   Card on Dark Web  
0                 0  
1                 0  
2                 0  
3                 0  
4                 0

In [43]:
df.to_csv('model_ready_dataset.csv')